src/cephadm/cephadm

```diff
@@ -2826,6 +2931,7 @@ def command_deploy():
     else:
         logger.info('%s daemon %s ...' % ('Deploy', args.name))
 
+    daemon_ports = []   # type: List[int]
     if daemon_type in Ceph.daemons:
```

Comment:
As you removed the firewall config from deploy_daemon, we have to set the firewall somewhere else.

what about the MON ports etc here? Are they missing?

what about the dashboard ports?

Reply:
I have removed firewall logic from deploy_daemon, because  i would like to convert this function in some moment in the future in something generic used to deploy a container in a host, using one container image, a config file, and a set of mounts.
This future function should be called from other more specific functions like "deploy_mon" , which should take care of specific things for a monitor daemon, as for eaxmple, once deployed the container open the right ports calling the generic Firewall object provided in this PR.

Port opening for monitor, manager, monitoring stack, dashboard are not lost. 
If you search `fw.open_ports` you will see the points where this happens. (most times replacing the rigid `update_firewall` function that I have removed)
Sorry, because I got details about this topic  in the manual tests I did and i forgot to include the information in the PR. ( i will update the PR with it)

Thought about "command_deploy":
In the same way that the best place to have the logic to deploy containers is "cephadm standalone", I think that it is also the right place to have all the firewall management stuff.
It would be nice to "work" always with "service specification" objects ( this objects must store all the information we know about one service) . And pass the reference to the spec object when it would be needed ( from cephadm to mgr/cephadm and reverse).
In mgr/cephadm we can check complex things like ports conflict, collocation, etc and validate the service spec object ...and in cephadm standalone we can just deploy the container using the information provided inside the validated service especification object.

Reply:
> I have removed firewall logic from deploy_daemon, because i would like to convert this function in some moment in the future in something generic used to deploy a container in a host, using one container image, a config file, and a set of mounts.
> This future function should be called from other more specific functions like "deploy_mon" , which should take care of specific things for a monitor daemon, as for eaxmple, once deployed the container open the right ports calling the generic Firewall object provided in this PR.

sounds good. Just be aware that `command_deploy` is the entrypoint when calling `cephadm deploy`

> 
> Port opening for monitor, manager, monitoring stack, dashboard are not lost.
> If you search `fw.open_ports` you will see the points where this happens. (most times replacing the rigid `update_firewall` function that I have removed)

Right. saw that.

* you open the MON ports in bootstrap
* you open the MGR ports in bootstrap
* you open the monitoring and NFS ports in `command_deploy`
* you open the monitoring ports in `command_adopt_{prometheus,grafana,alertmanager}`

what I'm missing is:

* Where do you open the ports for other Ceph daemons in `command_deploy`?


> Sorry, because I got details about this topic in the manual tests I did and i forgot to include the information in the PR. ( i will update the PR with it)

Saw that, thanks!

> Thought about "command_deploy":
> In the same way that the best place to have the logic to deploy containers is "cephadm standalone", I think that it is also the right place to have all the firewall management stuff.
> It would be nice to "work" always with "service specification" objects ( this objects must store all the information we know about one service) . And pass the reference to the spec object when it would be needed ( from cephadm to mgr/cephadm and reverse).
> In mgr/cephadm we can check complex things like ports conflict, collocation, etc and validate the service spec object ...and in cephadm standalone we can just deploy the container using the information provided inside the validated service especification object.

sounds great! Just be aware that we have some uncanny compatibility issues between bin/cephadm and mgr/cephadm:

* tasks/cephadm.py in master needs to be compatible to src/cephadm in v15.2.0
* src/cephadm in the octopus branch (and also in master!) needs to be compatible to the latest octopus release

I don't want to start with the compatibility issues of the  `packaged` mode, as this will be really bad. Like:

* Pacific mgr/cephadm will call an octopus bin/cephadm (-2 major versions)

Reply:
It will be acceptable to start the migration to a more friendly system?

we will need to do: 
in src/cephadm:
1. "cephadm deploy" should accept a new parameter  "daemon_spec"  which will include the things we need to run a container. 

- We will start writing in this spec (yaml string) only what we need to manage ports. (firewall service and ports)
In the case of the manager, for example, we can provide only the ports really used ( dashboard + prometheus exporter) , taking into account any kind of user customization.

- We can continue making more rich the spec parameter content, with other properties. The final target will be to create daemons using only this parameter.

in cephadm orchestrator:
2. Add functionality to  "_create_daemon" to deal with the different kind of daemons we can create, in order to generate the right "daemon_spec" content to use in the "cephadm deploy" command.

Reply:
> It will be acceptable to start the migration to a more friendly system?

I guess we have to solve the packaged cephadm mode, before we can start messing with the CLI. 

> 
> we will need to do:
> in src/cephadm:
> 
> 1. "cephadm deploy" should accept a new parameter  "daemon_spec"  which will include the things we need to run a container.
> 
> * We will start writing in this spec (yaml string) only what we need to manage ports. (firewall service and ports)
>   In the case of the manager, for example, we can provide only the ports really used ( dashboard + prometheus exporter) , taking into account any kind of user customization.
> * We can continue making more rich the spec parameter content, with other properties. The final target will be to create daemons using only this parameter.

Note that the [deploy](https://docs.ceph.com/docs/master/man/8/cephadm/#deploy) command CLI is already pretty generic. Maybe we can simply enhance it?

> 
> in cephadm orchestrator:
> 2. Add functionality to "_create_daemon" to deal with the different kind of daemons we can create, in order to generate the right "daemon_spec" content to use in the "cephadm deploy" command.

you mean as in #35471 ?

Reply:
@sebastian-philipp : 
In the "command_deploy" method there is a call at the end to open the ports and create the services with the firewall.

But in any case .... in my view this is wrong because:
- The list of ports/services to open in the firewall should be provided as an specific parameter, or as part of a well defined structure ( the daemon specification) used as parameter in this method . Now we have a generic "config-json" where we can put this information ... but I do not like to use this "bag" to pass critical settings.
- In my view the "command_deploy" should have all the things needed to deploy a container and do not need to do specific actions depending of the type of daemon.
- In the cephadm orchestrator we are using "harcoded" values for ports in several services. To change this implies to create this services using a "service specification", that will generate a "daemon specification" (not as in  #35471, because the things i expect to have in a daemon specification are things like, image, ports, fw_service, mounts, config files, placement ...)
- We will need more changes in cephadm orchestrator services to manage properly service specification provided by users.

This means to generate another huge pull request that probably will be rejected because the amount of code to review, the high risk implied and because it will be very difficult to preview all the collateral effects...

So it seems that the only possibility is to add in the command deploy more specific stuff to deal with the ports of the different kind of daemons.
I think that in this moment it has exactly the same port coverage we had previously. Can you see any gap?

Reply:
> I think that in this moment it has exactly the same port coverage we had previously. Can you see any gap?

Yes, Im missing the place where you open the ports for other Ceph daemons in command_deploy. `daemon_ports` is never set within the `if daemon_type in Ceph.daemons` block.

Reply:
> > I think that in this moment it has exactly the same port coverage we had previously. Can you see any gap?
> 
> Yes, Im missing the place where you open the ports for other Ceph daemons in command_deploy. `daemon_ports` is never set within the `if daemon_type in Ceph.daemons` block.

In the case that the daemon_type will be inside the Ceph.daemons, previously we had that the <create_daemon> function calling the "update_firewall" function:

And in this function what we had was:
 - a fw-service added to the firewall if the daemon_type was in (mon, mgr, mds, osd)
 
 We have now exactly the same functionality, but now it is placed in the call to:

` fw.enable_service_for(daemon_type)  At the end of "command_deploy"`

 - A set of ports added if the daemon_type was part of the monitoring_stack daemons or NFSGanesha:

 we have exactly the same functionality in "command_deploy". In fact I didn't touch nothing here. 
 The ports for the monitoring stack daemons or the NFSGanesha daemons are set explicitly into the
 "daemon_ports" vector

 - A set of ports added for the manager.
 Two of this ports were for the dashboard and one for the prometheus node exporter. 
 Cephadm orchestrator does not have the capability to manage the dashboard or the prometheus node exporter, 
 so the ports to open depends if these manager modules (dashboard, prometheus are enabled or not, and the current configuration)
 So the code inside the "update_firewall" function was not correct because it opened always the same ports without taking into account that these ports could be customized.

In order to solve this problem, I have included  the detection of the ports used by manager modules, and passed this list as part of the extra-args param in the cephadm deploy command.